src/miner.cpp

```diff
@@ -380,19 +375,17 @@ bool static ScanHash(const CBlockHeader *pblock, uint32_t& nNonce, uint256 *phas
     hasher.Write((unsigned char*)&ss[0], 76);
 
     while (true) {
-        nNonce++;
+        pblock->nNonce++;
 
         // Write the last 4 bytes of the block header (the nonce) to a copy of
         // the double-SHA256 state, and compute the result.
-        CHash256(hasher).Write((unsigned char*)&nNonce, 4).Finalize((unsigned char*)phash);
+        CHash256(hasher).Write((unsigned char*)&pblock->nNonce, 4).Finalize((unsigned char*)&hash);
 
-        // Return the nonce if the hash has at least some zero bits,
-        // caller will check if it has enough to reach the target
-        if (((uint16_t*)phash)[15] == 0)
+        if (CheckProofOfWork(hash, pblock->nBits, chainparams))
```

Comment:
No, CheckProofOfWork is not much more expensive than ```((uint16_t*)phash)[15] == 0 && UintToArith256(hash) <= arith_uint256().SetCompact(pblock->nBits)```, which are the checks that CheckProofOfWork is replacing.
In fact, in regtest mode (the mining we care about the most, right?) those checks didn't existed and CheckProofOfWork was already being called.

I also have commits to decouple the pow functions from util that would remove the logging of the errors in the internal checks, but I would really like to wait for  #6051 (or at least #5975 and https://github.com/jtimon/bitcoin/commit/543f5b2ba0769f9e54cb667efbf74e040ea3119d ) to be merged before proposing that.